### DISCUSSION TITLE
Double free error in tcache 2

### DISCUSSION THREAD

**Discussion by rharkor** [fastapi/fastapi#6324] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
for (let i = 0; i < 2; i++) {
  axios
    .post(URL + '/data/fetch/salle', getContent())
    .then(res => {
      console.log(`statusCode: ${res.status}`)
      // console.log(res)
    })
    .catch(error => {
      // console.error(error)
      console.error("Error")
    })
}
```

### Description

Hi, I wanted to do a stress test on my api and make a lot of request to see if it can handle it.
I precise my API run in a docker container with a flask website (with nginx) and a database (mysql) each request call the database to execute a select.

The problem is that my api is working well but when I do plenty of simultaneous request it crash with the error double free or either, free(): error in tcache 2. Can someone help me please ?

secondtour-api      | INFO:     192.168.224.1:45528 - "POST /data/fetch/salle HTTP/1.1" 200 OK
secondtour-api      | INFO:     192.168.224.1:45530 - "POST /data/fetch/salle HTTP/1.1" 200 OK
secondtour-api      | INFO:     192.168.224.1:45532 - "POST /data/fetch/salle HTTP/1.1" 200 OK
secondtour-api      | double free or corruption (fasttop)
secondtour-api exited with code 139

### Operating System

Linux, Windows

### Operating System Details

I'm on windows and the api run in a linux container

### FastAPI Version

0.73.0

### Python Version

Python 3.9.5

**Comment by rharkor** [comment]
Okay, I find that it's related to my database mysql, i'll dig in to resolve it. If you have any idea ?

**Comment by rharkor** [comment]
Okay I still can't find, I also get this error sometimes "free(): double free detected in tcache 2". Finally it doesnt seem to be an sql error. I try it in an isolated environment then the results are really strange.. In python I don't have any problem even if I spam the request (code below) : 
```py
    import asyncio
    loop = asyncio.get_event_loop()

    async def do_request():
        response =  requests.post("http://127.0.0.1:443/data/fetch/salle", json={"username": "admin", "password": "L0calAdmin"})
        print(response.json())
    async_tasks = []
    for i in range(100):
        async_tasks.append(loop.create_task(do_request()))
    print("Send !")

    loop.run_until_complete(asyncio.gather(*async_tasks))
```

But the same request in js make the console crash ! 
```js

for (let i = 0; i < 50; i++) {
  axios
    .post(URL + '/data/fetch/salle', getContent())
    .then(res => {
      console.log(`statusCode: ${res.status}`)
      // console.log(res)
    })
    .catch(error => {
      // console.error(error)
      console.error("Error")
    })
}
```

Please help...

**Comment by jgould22** [comment]
Without seeing server side code it is very difficult to help you.

To better help you need to provide us with at least a small FastAPI code sample that produces the issue.

Additionally since `requests` is a sync lib I think your python code is only going to send one request at a time which is why you might not see the behavior with that python snippet.

**Comment by rharkor** [comment]
> Without seeing server side code it is very difficult to help you.
> 
> To better help you need to provide us with at least a small FastAPI code sample that produces the issue.
> 
> Additionally since `requests` is a sync lib I think your python code is only going to send one request at a time which is why you might not see the behavior with that python snippet.

Hi, to do some test I create a version of the api with one single route and all in the same file, this also don't work when I spam the request at the same time.


```py
import datetime
import string
import json
from time import sleep
import traceback
import requests
import mysql.connector
from dotenv import load_dotenv
load_dotenv()

class MySQLDatabase:
    def __init__(self) -> None:
        self.db = None
    
    def connect(self, _host, _user, _password, _database, _it=0):
        try:
            self.db = mysql.connector.connect(
            host=_host,
            user=_user,
            password=_password,
            database=_database,
            autocommit=True
            )
            print("Connected !")
        except mysql.connector.errors.DatabaseError:
            if _it < 10:
                traceback.print_exc()
                print("Can't connect to database, retry in 10seconds...", _host, _user, _password, _database)
                sleep(10)
                self.connect(_host, _user, _password, _database, _it)
            else:
                print("Exiting!")
                return
        
    def myconverter(self, o):
        if isinstance(o, datetime.datetime):
            return o.strftime("%a %b %d %H:%M:%S %Y")
        
    def query(self, _query: string):
        if self.db:
            mycursor = self.db.cursor(dictionary=True, buffered=True)
            mycursor.execute(_query)
            if "SELECT" in _query or "SHOW" in _query:
                myresult = mycursor.fetchall()
                output = json.loads(json.dumps(list(myresult), default = self.myconverter))
            else:
                myresult = mycursor.lastrowid
                output = json.loads(json.dumps({"id": myresult}, default = self.myconverter))
            mycursor.close()
            return output
        else:
            raise Exception('Please connect the db first')


import os

DB_HOST = '127.0.0.1'
DB_USER = 'api'
DB_PWD = 'api'
DB_NAME = 'secondtour'

db = MySQLDatabase()
db.connect(DB_HOST, DB_USER, DB_PWD, DB_NAME)


from fastapi import FastAPI, status
from fastapi.responses import JSONResponse


app = FastAPI()

sensible_tables = os.getenv('SENSIBLE_TABLES').split(",")


def test_connection(user):
    if os.getenv('AUTH_USERNAME') == user["username"] and os.getenv('AUTH_PASSWORD') == user["password"]:
        return True
    return False

@app.post("/data/fetch/salle")
def read_root():
    if not test_connection({"username": "admin", "password": "L0calAdmin"}):
        output = JSONResponse(status_code=status.HTTP_400_BAD_REQUEST, content={'error': "Incorrect identifiers"})
    elif "salle" in (item.lower() for item in sensible_tables):
        output = JSONResponse(status_code=status.HTTP_401_UNAUTHORIZED ,content={'error': "This table cannot be accessed due to his privacy settings"})
    else:
        try:
            output = JSONResponse(status_code=status.HTTP_200_OK, content=db.query(f"SELECT * FROM SALLE;"))
        except mysql.connector.errors.ProgrammingError:
            output = JSONResponse(status_code=status.HTTP_404_NOT_FOUND ,content={'error': "Error while fetching data"})
        except Exception:
            output = JSONResponse(status_code=status.HTTP_404_NOT_FOUND ,content={'error': traceback.format_exc()})
    return output


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
``` 
main.py

This is the code to do all the requests

```js
require('dotenv').config()
const axios = require('axios')

function getContent (content = {}) {
  return {
    username: process.env.API_USERNAME,
    password: process.env.API_PASSWORD,
    content: content
  }
}

const URL = process.env.API_URL

for (let i = 0; i < 100; i++) {
  axios
    .post(URL + '/data/fetch/salle', getContent())
    .then(res => {
      console.log(`statusCode: ${res.status}`)
      // console.log(res)
    })
    .catch(error => {
      // console.error(error)
      console.error("Error")
    })
}
```
manual_test.js

Thank you for your reply

**Comment by jgould22** [comment]
Right, it seems likely that you cannot share that db connection between tasks or threads as it might not thread safe.

I would first try moving your db connection into your `read_root` function and then see if that causes the issue.

This will be slower(since it is initializing a new db connection for each request)  but if you don't get there error anymore you know that will be the issue.

Then you will need to refactor your application so it is not sharing that db (and probably change the MySQL library you are using to something that is async, perhaps aiomysql).
